### PR TITLE
[TT-XLA][PyTorch] Llama-3.2-1B Full Fine-Tuning

### PR DESCRIPTION
**Pull request by mcupacTT**
**Issue**

#515 

**Training Workload Setting**

- Model: Llama-3.2-1B
- Strategy: Full fine-tuning (SFT)
- Framework: PyTorch
- Dataset: Metamathqa
- Hardware: Blackhole p150

**What's Changed**

This PR introduces full fine-tuning of Llama-3.2-1B on p150.
The PR implements scheduler logic in the train script.

**Experiment Results**

<table>
  <tr>
     <td>Train Loss
     <td>Val Loss
  <tr>
  <tr>
    <td>W B Chart 6_4_2026, 5_41_51 PM</td>
    <td>W B Chart 6_4_2026, 5_41_57 PM</td>
  </tr>
</table>

- GPU run [wandb](https://wandb.ai/tenstorrent-ml/llama1b-singlechip-full/runs/4ai55dxf?nw=nwusermcupac)
- TT run [wandb](https://wandb.ai/tenstorrent-ml/llama1b-singlechip-full/runs/o0o1hgai?nw=nwusermcupac)

**Remarks**
- The training was performed on 10% of randomly shuffled dataset using the standard seed from the configuration - this is done to accelerate initial dataset mapping and filtering, as well as the time required for validation pass - these changes are not included in the PR so that the original metamathqa configuration is not altered
- The first training iteration is investigated with chisel - on fwd+bw graph and on optimizer graph - both investigations showed **no issues in PCC values** (isolated and accumulated)

**Checklist**

- [x]      Addition of appropriate .yaml file
- [x]      TT device testing
- [x]      GPU testing

### CODE REVIEW

File: blacksmith/experiments/torch/llama/xla/full/llama_3_2_1b_alpaca.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ignored_index: -100
+
+# Reproducibility settings
+seed: 23
+deterministic: False
+
+# Logging settings
+log_level: "INFO"
+use_wandb: True
+wandb_project: "llama1b-singlechip-full"
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/full/llama_3_2_1b_alpaca.yaml:25]:
```suggestion
wandb_project: "llama1b-singlechip-full-sft"
```
nit

File: blacksmith/models/torch/huggingface/hf_models.py [hunk 2]
```diff
@@ -36,11 +36,14 @@ def get_model(config: TrainingConfig, device: torch.device):
...
     # Apply training specific modifications
-    # Apply LoRA if rank is specified
     if config.training_type == "lora":
         model = _apply_lora(model, config)
     elif config.training_type == "adapters":
         _apply_adapters(model, config)
+    elif config.training_type == "full":
+        # Full fine-tuning: leave the base model untouched so every parameter
+        # remains trainable (requires_grad=True from HF defaults).
+        pass
```
Comment (mmilosevicTT) [inline blacksmith/models/torch/huggingface/hf_models.py:46]: Is this true always? HF can freeze some parameters. Do we want to make sure all of them are unfrozen?

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 3]
```diff
@@ -126,14 +138,31 @@ def train(
...
+            optimizer,
+            schedulers=[
+                LinearLR(optimizer, start_factor=1e-2, total_iters=warmup_steps),
+                CosineAnnealingLR(optimizer, T_max=total_steps - warmup_steps, eta_min=config.learning_rate * 0.1),
+            ],
+            milestones=[warmup_steps],
+        )
+    else:
+        scheduler = None
+
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py:165]: Scheduler setting logic should not be in a training script. It should be a tool extracted that just returns the scheduler.

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 4]
```diff
@@ -126,14 +138,31 @@ def train(
...
-    logger.info(f"Loaded {config.dataset_id} dataset. Eval dataset size: {len(eval_dataloader)*config.batch_size}")
+    logger.info(f"Loaded {config.dataset_id} dataset. Eval dataset size: {len(eval_dataloader) * config.batch_size}")
 
     tokenizer = train_dataset.tokenizer
 
+    # Cosine annealing with linear warmup.
+    total_steps = (
+        config.total_steps or (len(train_dataloader) // config.gradient_accumulation_steps) * config.num_epochs
+    )
+    warmup_steps = config.warmup_steps or max(1, int(total_steps * 0.03))
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py:153]: Do we really not want to give option for no warmup?

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 5]
```diff
@@ -186,14 +215,22 @@ def train(
 
                 # Only step the optimizer after accumulating gradients.
                 if accumulation_step == config.gradient_accumulation_steps:
+
                     device_manager.optimizer_step(optimizer)
+                    if scheduler is not None:
+                        scheduler.step()
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py:221]: Do we need a sync after this?

**Review by mmilosevicTT — COMMENTED** [review]
Left some comments. 

There are several other things we should probably do:
- Please link the issue that fixes the PCC problems we had.
- Please add all unrelated changes to PR description (addition of lr scheduling, addition of decode loop in train.py for llama, logging improvements...)
